etk-asm/src/asm.rs

```diff
@@ -430,202 +311,137 @@ impl Assembler {
                     }
+                    Err(ops::Error::ContextIncomplete {
+                        source: UnknownMacro { name, .. },
+                    }) => return error::UndeclaredInstructionMacro { name }.fail(),
+                    Err(ops::Error::ContextIncomplete {
+                        source: UndefinedVariable { name, .. },
+                    }) => return error::UndeclaredVariableMacro { var: name }.fail(),
                 }
-
-                Ok(())
             }
             RawOp::Raw(raw) => {
                 self.concrete_len += raw.len();
-                self.ready.extend(raw);
-                Ok(())
+                self.ready.push(RawOp::Raw(raw.to_vec()));
+            }
+            RawOp::Scope(scope) => {
+                let mut asm = Self::new();
+                let scope_result = asm.assemble(&scope)?;
+                self.concrete_len += scope_result.len();
+                self.ready.push(RawOp::Raw(scope_result));
             }
         }
+
+        Ok(self.concrete_len)
     }
 
-    fn push_pending(&mut self, rop: RawOp) -> Result<(), Error> {
-        // Update total size of pending ops.
-        if let Some(ref mut pending_len) = self.pending_len {
-            match rop.size() {
-                Some(size) => *pending_len += size,
-                None => self.pending_len = None,
+    /// Backpatch dynamic operations and emit the assembled program.
+    ///
+    /// Errors if there are any undeclared labels.
+    fn backpatch_and_emit(&mut self) -> Result<Vec<u8>, Error> {
+        if !self.undeclared_labels.is_empty() {
+            return error::UndeclaredLabels {
+                labels: self
+                    .undeclared_labels
+                    .iter()
+                    .map(|l| l.to_owned())
+                    .collect::<Vec<String>>(),
             }
+            .fail();
         }
+        let mut output = Vec::new();
+        for op in self.ready.iter() {
+            if let RawOp::Op(ref op) = op {
+                match op
+                    .clone()
+                    .concretize((&self.declared_labels, &self.declared_macros).into())
+                {
+                    Ok(mut cop) => {
+                        if let AbstractOp::Push(imm) = op {
+                            let exp = imm.tree.eval_with_context(
+                                (&self.declared_labels, &self.declared_macros).into(),
+                            );
+
+                            let labels: HashSet<String> = imm
+                                .tree
+                                .labels(&self.declared_macros)
+                                .unwrap()
+                                .into_iter()
+                                .collect();
+                            if !labels.is_empty() {
+                                if let Ok(val) = exp {
+                                    let extra = val.to_usize().unwrap() / 256;
```

Comment:
Hi Sam. I improved a little the documentation of the function and added some comments so that you can better understand what's happening in each step. I hope it's clearer now :)